File: solutions/python/dnd-character/1/dnd_character.py
import random


def modifier(score):
    return (score - 10) // 2


class Character:
    def __init__(self):
        self.strength = self.ability()
        self.dexterity = self.ability()
        self.constitution = self.ability()
        self.intelligence = self.ability()
        self.wisdom = self.ability()
        self.charisma = self.ability()
        self.hitpoints = 10 + modifier(self.constitution)

    @staticmethod
    def ability():
        rolls = sorted([random.randint(1, 6) for _ in range(4)])
        return sum(rolls[1:])

File: solutions/ruby/dnd-character/5/dnd_character.rb
=begin
Write your code for the 'D&D Character' exercise in this file. Make the tests in
`dnd_character_test.rb` pass.

To get started with TDD, see the `README.md` file in your
`ruby/dnd-character` directory.
=end

class DndCharacter
  attr_reader :strength, :dexterity, :constitution, :intelligence, :wisdom, :charisma, :hitpoints

  def self.modifier(score)
    ((score - 10) / 2.0).floor
  end

  def self.ability
    4.times.map { rand(1..6) }.sort.drop(1).sum
  end

  def initialize
    @strength = self.class.ability
    @dexterity = self.class.ability
    @constitution = self.class.ability
    @intelligence = self.class.ability
    @wisdom = self.class.ability
    @charisma = self.class.ability
    @hitpoints = 10 + self.class.modifier(@constitution)
  end
end
